### PR TITLE
Temporal module;

### PR DESCRIPTION
**Pull request by cityremade**
Date formatting functions placed into one object within mapp.utils.temporal{}.

Previously they were only available inside date/datetime location entries and inaccessible outside.
With date and times stored as unix epoch they can be accessed as utilities in custom scripts/views/plugins.

Adds function which converts date string to unix epoch which previously wasn't reusable.

## GitHub Issue
https://github.com/GEOLYTIX/xyz/issues/1694

Tested locally.
Non-breaking change as these methods were not used outside date/time entry module.
Updated documentation on date formatting;

All date and datetime read only and editable entries should receive correct formatting
in location view and editable date inputs.

### CODE REVIEW

**Comment by simon-leech** [comment]
@cityremade Can you change the base branch - this shouldn't be main. It should be patch (v4.12.4) or minor (v4.13.0)? 😅

**Review by dbauszus-glx — CHANGES_REQUESTED** [review]
I have updated the docs for the entry/date module.

I also added the properties for the params in the temporal utils module.

I don't fully understand the `fr-CA` default? Should this not be english as default derived from the user language.

File: lib/utils/temporal.mjs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+Returns time only as 00:00.
+
+@param {infoj-entry} entry type:date or type:datetime infoj-entry typedef object.
+@property {integer} entry.value date[time] as unix epoch.
+@property {integer} [entry.newValue] updated entry.value which has not yet been saved.
+@returns {String}
+*/
+function timeString(entry) {
+  return new Date((entry.newValue || entry.value) * 1000)
+    .toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' })
```
Comment (RobAndrewHurst) [inline lib/utils/temporal.mjs:47]: image

we are going to want to use the [Intl.DateTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat) object rather as this is going to be way more efficient.

**Review by RobAndrewHurst — CHANGES_REQUESTED** [review]

**Comment by cityremade** [comment]
In order to set the date on date input we need it YYYY-MM-DD format hence `fr-CA`.
These methods make it possible to set current value on input of type date, otherwise we get empty fields.

File: lib/utils/temporal.mjs (hunk 1 above)
Comment (cityremade) [reply lib/utils/temporal.mjs:47]: does it mean formatting by toLocaleString should be replaced entirely?

Comment (RobAndrewHurst) [reply lib/utils/temporal.mjs:47]: Yeah! We did the same for the numeric formatter

**Comment by cityremade** [comment]
String date formatting now handled by `Intl.DateTimeFormat`
Date/date time inputs now use only naive ISO Date derived from unix timestamp.
timeString function removed.
All dates are handled by one function dateString and custom values will be applied in entry configuration according to Js docs ( entry.locale + entry.options{} )

```
new Intl.DateTimeFormat('en-GB', {
    dateStyle: 'full',
    timeStyle: 'long',
    timeZone: 'Australia/Sydney',
  }).format(date),
);
// Expected output: "Sunday, 20 December 2020 at 14:23:16 GMT+11"
```

**Comment by RobAndrewHurst** [comment]
I renamed the module 'temporal' as it was a bit vague to the layman of what the module is actually doing. So I gave it a clearer name of `dateTime`

**Comment by RobAndrewHurst** [comment]
I have also update the object name passed into the function on the utility. It was specifying that the object was only going to come from infoj entries which shouldn't be true for a utility that can be used by other elements potentially. So I have made the object passed in a more generic approach that can apply to more situations than just infoj entries.

**Review by RobAndrewHurst — APPROVED** [review]
I am all happy with this!

I have added in tests to check each function.

**Comment by dbauszus-glx** [comment]
The dateTime utils referenced in the date entry module needed changing to temporal.

**Review by dbauszus-glx — APPROVED** [review]
Works well. Only change I did was to take the language as locale from the mapp.user if not explicit on entry [params].

This shows the correct formatting with dots for german [de] locale.

image

**Pull request merged**